test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestFeatureService.java

```diff
@@ -10,57 +10,136 @@
-        ).orElse(featureId -> true); // We can safely assume that new non-semantic versions (serverless) support all historical features
+    ESRestTestFeatureService(List<FeatureSpecification> featureSpecs, Collection<Version> nodeVersions, Set<String> clusterStateFeatures) {
+        List<FeatureSpecification> specs = new ArrayList<>(featureSpecs);
+        specs.add(new RestTestLegacyFeatures());
+        if (MetadataHolder.HISTORICAL_FEATURES != null) {
+            specs.add(MetadataHolder.HISTORICAL_FEATURES);
+        }
+        var historicalFeatures = FeatureData.createFromSpecifications(specs).getHistoricalFeatures();
+        this.knownHistoricalFeatureNames = Optional.ofNullable(historicalFeatures.lastEntry()).map(Map.Entry::getValue).orElse(emptySet());
         this.clusterStateFeatures = clusterStateFeatures;
+        this.allSupportedFeatures = Sets.union(
+            clusterStateFeatures,
+            nodeVersions.stream()
+                .min(Comparator.naturalOrder())
+                .map(v -> Optional.ofNullable(historicalFeatures.floorEntry(v)).map(Map.Entry::getValue).orElse(emptySet()))
+                .orElse(knownHistoricalFeatureNames)
+        );
     }
 
-    private static boolean hasHistoricalFeature(NavigableMap<Version, Set<String>> historicalFeatures, Version version, String featureId) {
-        var features = historicalFeatures.floorEntry(version);
-        return features != null && features.getValue().contains(featureId);
+    public static boolean hasFeatureMetadata() {
+        return MetadataHolder.HISTORICAL_FEATURES != null;
     }
 
     @Override
     public boolean clusterHasFeature(String featureId) {
-        if (clusterStateFeatures.contains(featureId)) {
-            return true;
+        if (hasFeatureMetadata()
+            && MetadataHolder.FEATURE_NAMES.contains(featureId) == false
+            && knownHistoricalFeatureNames.contains(featureId) == false) {
+            throw new IllegalArgumentException(
+                Strings.format(
+                    "Unknown feature %s: check the feature has been added to the correct FeatureSpecification in the relevant module or, "
+                        + "if this is a legacy feature used only in tests, to a test-only FeatureSpecification such as %s.",
+                    featureId,
+                    RestTestLegacyFeatures.class.getCanonicalName()
+                )
+            );
         }
-        return historicalFeaturesPredicate.test(featureId);
+        if (MetadataHolder.FEATURE_NAMES.contains(featureId)) {
+            return clusterStateFeatures.contains(featureId);
+        }
+        return allSupportedFeatures.contains(featureId);
     }
 
     @Override
     public Set<String> getAllSupportedFeatures() {
         return allSupportedFeatures;
     }
+
+    private static class MetadataHolder {
+        private record HistoricalFeatureSpec(Map<NodeFeature, Version> historicalFeatures) implements FeatureSpecification {
```

Comment:
This could just be an anonymous class, returning `historicalFeatures` captured in a closure